src/emulator/auth/state.ts

```diff
@@ -53,14 +54,21 @@ export abstract class ProjectState {
 
   abstract get enableEmailLinkSignin(): boolean;
 
-  createUser(props: Omit<UserInfo, "localId" | "createdAt" | "lastRefreshAt">): UserInfo {
+  abstract shouldForwardCredentialToBlockingFunction(
+    type: "accessToken" | "idToken" | "refreshToken"
+  ): boolean;
+
+  abstract getBlockingFunctionUri(event: BlockingFunctionEvents): string | undefined;
+
+  generateLocalId(): string {
     for (let i = 0; i < 10; i++) {
       // Try this for 10 times to prevent ID collision (since our RNG is
       // Math.random() which isn't really that great).
       const localId = randomId(28);
-      const user = this.createUserWithLocalId(localId, props);
-      if (user) {
-        return user;
+      if (!this.users.has(localId) && !this.pendingLocalIds.has(localId)) {
+        // Create a pending localId until user is created
+        this.pendingLocalIds.add(localId);
```

Comment:
This creates a memory leak if the blocking functions throws and the `localId` is never used. (Theoretically, if this happens enough times we will even run out of IDs.) I think this is probably acceptable, but worth a comment to document the tradeoff.